### PR TITLE
test: Added Unit tests for stream.go file's methods

### LINKED ISSUE #661: feat: Add Unit Tests for stream.go
#### **Problem**

The OLake codebase currently has unit tests for:

* `resolver`
* `flattener`
* `catalog`

However, the file `olake/types/stream.go` does not have any test coverage.

This file handles critical stream definition and configuration logic for the OLake connector.

Without unit tests, bugs in stream construction, incorrect destination name generation, builder method chaining issues, schema field management, JSON deserialization edge cases, or utility function failures may slip through during development. This is particularly critical as the `Stream` struct is used extensively across the codebase.

---

#### **Proposed Solution**

Add comprehensive unit tests for the logic inside `stream.go`.

This will help:

* Increase overall test coverage
* Make contributions safer when modifying stream-related logic

For reference on how existing unit tests are structured, you can refer to **PR #273** and the existing `catalog_test.go` file.

---

#### **Notes for Contributors**

* Add tests covering most of the scenarios and edge cases which should indicate both pass and fail scenarios.

* Follow the existing test file structure and naming conventions to maintain consistency with other parts (see `catalog_test.go` for reference).

* Use `testify/assert` and `testify/require` for assertions (as seen in existing test files).

### PR DESCRIPTION
**Pull request by supersanchayrx**
# Description

Added comprehensive unit tests for `stream.go` file.

Fixes #661 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

All tests verified via `go test ./types -v -run "TestStream_"`
- [x] **TestStream_NewStream** — 8 table-driven cases: empty name, empty namespace, nil/empty sourceDatabase, all empty, all filled, special characters (Reformat verification), viper prefix.
- [x] **TestStream_ID** — 5 table-driven cases: empty name, empty namespace, all filled, all empty, special characters (no reformat applied).
- [x] **TestStream_WithSyncMode** — 5 table-driven cases: single, multiple, all, duplicate, empty modes.
- [x] **TestStream_WithPrimaryKey** — 6 table-driven cases: single, multiple, composite, duplicate, empty, underscore keys.
- [x] **TestStream_WithCursorField** — 6 table-driven cases: single, multiple, timestamp fields, duplicate, empty, underscore fields.
- [x] **TestStream_WithSchema** — 2 subtests: set new schema, replace existing schema (pointer verification).
- [x] **TestStream_Wrap** — 4 table-driven cases: index 0, 1, negative, large index.
- [x] **TestStream_WithUpsertField** — 11 table-driven cases (nullable/non-nullable string, int, float, bool, timestamp + olake column + special char column) + merge subtest verifying type accumulation on same column.
- [x] **TestStream_UnmarshalJSON** — 5 subtests: safe initialization on missing fields, correct data loading, invalid JSON (error path), empty JSON (set initialization), all fields populated.
- [x] **TestStream_StreamsToMap** — 4 subtests: empty input, single stream, multiple streams, duplicate IDs (last wins).
- [x] **TestStream_LogCatalog** — Verifies catalog file creation, valid JSON output, and correct stream count.

# Screenshots or Recordings

https://github.com/user-attachments/assets/79caa5da-52a1-4f99-8b66-e54485d66478

# Video Content :
All unit test cases pass.

https://github.com/user-attachments/assets/7ac6a4e1-ee6c-42e1-b0d4-1dd2fff3f6e1

# Video Content: 
The streams.go interaction with other files still work properly as the sync shows expected behaviour.

## Documentation

- [x] N/A (bug fix, refactor, or test changes only)

## Related PR's (If Any):
#680

### CODE REVIEW

**Comment by supersanchayrx** [comment]
Also very grateful to @Beeram12 for his contributions. These tests were added on top of his work done from pr #680

**Comment by nayanj98** [comment]
@supersanchayrx Thanks a lot! Will assign a reviewer for this PR soon.

**Comment by nayanj98** [comment]
@supersanchayrx Assigning @saksham-datazip to review your PR

File: types/stream_test.go [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+	"testing"
+
+	"github.com/datazip-inc/olake/constants"
+	"github.com/datazip-inc/olake/utils"
+	"github.com/goccy/go-json"
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestStream_NewStream(t *testing.T) {
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can you change this name to TestNewStream following OLAKE format and also make changes in other functions where it is needed.

Comment (saksham-datazip) [inline types/stream_test.go]: Also please resolve these errors.

Image

File: types/stream_test.go [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+		name           string
+		namespace      string
+		sourceDatabase *string
+		viperPrefix    string
+
+		//expected values
+		expectedName      string
+		expectedNamespace string
+		expectedDatabase  string
+		expectedTable     string
```
Comment (saksham-datazip) [inline types/stream_test.go]: We can define an expectedStream (*Stream) and compare it directly against the result. That would make the test cleaner and remove the need for these four separate expected variables.

File: types/stream_test.go [hunk 3]
```diff
@@ -0,0 +1,793 @@
...
+
+	tests := []struct {
+		//test identifier
+		testName string
+
+		//test values
+		name           string
+		namespace      string
+		sourceDatabase *string
+		viperPrefix    string
```
Comment (saksham-datazip) [inline types/stream_test.go]: We are creating unit Tests not olake process so ig we dont need to test this viperPrefix path as inside the function it is not handling this case specially.

File: types/stream_test.go [hunk 4]
```diff
@@ -0,0 +1,793 @@
...
+			namespace:      "My.Schema",
+			sourceDatabase: nil,
+
+			expectedName:      "User-Orders.v2",
+			expectedNamespace: "My.Schema",
+			expectedDatabase:  ":my_schema",
+			expectedTable:     "user_orders_v2",
+		},
+		{
+			testName:       "stream with a custom prefix",
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can we remove this testCase As discussed earlier?

File: types/stream_test.go [hunk 5]
```diff
@@ -0,0 +1,793 @@
...
+			stream := NewStream(tt.name, tt.namespace, tt.sourceDatabase)
+
+			asserts.Equal(tt.expectedName, stream.Name)
+			asserts.Equal(tt.expectedNamespace, stream.Namespace)
+			asserts.NotNil(stream.SupportedSyncModes, "SupportedSyncModes should be initialized")
+			asserts.NotNil(stream.SourceDefinedPrimaryKey, "SourceDefinedPrimaryKey should be initialized")
+			asserts.NotNil(stream.AvailableCursorFields, "AvailableCursorFields should be initialized")
+			asserts.NotNil(stream.Schema, "Schema should be initialized")
+			asserts.Equal(tt.expectedDatabase, stream.DestinationDatabase)
+			asserts.Equal(tt.expectedTable, stream.DestinationTable)
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can you check if we can reduce these multiple checks?

File: types/stream_test.go [hunk 6]
```diff
@@ -0,0 +1,793 @@
...
+		})
+	}
+}
+
+func TestStream_ID(t *testing.T) {
```
Comment (saksham-datazip) [inline types/stream_test.go]: Please change name as told above you can take reference from reformat.go?

File: types/stream_test.go [hunk 7]
```diff
@@ -0,0 +1,793 @@
...
+	tests := []struct {
+		testName   string
+		name       string
+		namespace  string
+		expectedID string
+	}{
+		{
+			testName:   "name field empty but all other filled",
```
Comment (saksham-datazip) [inline types/stream_test.go]: no need to say all other field as we just have two fields we can directly mention it here also make the respective changes wherever possible.

File: types/stream_test.go [hunk 8]
```diff
@@ -0,0 +1,793 @@
...
+			assert.Equal(t, tt.expectedID, stream.ID())
+		})
+	}
+}
+
+func stringPtr(s string) *string {
+	return &s
+}
+
+func TestStream_WithSyncMode(t *testing.T) {
```
Comment (saksham-datazip) [inline types/stream_test.go]: Change name

File: types/stream_test.go [hunk 9]
```diff
@@ -0,0 +1,793 @@
...
+	tests := []struct {
+		name             string
+		modes            []SyncMode
+		expectedModes    []SyncMode
+		notExpectedModes []SyncMode
```
Comment (saksham-datazip) [inline types/stream_test.go]: No need for notExpectedModes variable in tests struct ?

File: types/stream_test.go [hunk 10]
```diff
@@ -0,0 +1,793 @@
...
+
+			// check if the set contains added modes
+			for _, mode := range tt.expectedModes {
+				assert.True(t, outputStream.SupportedSyncModes.Exists(mode), "Should contain %v", mode)
+			}
+
+			// check if the set does not contain other modes
+			for _, mode := range tt.notExpectedModes {
+				assert.False(t, outputStream.SupportedSyncModes.Exists(mode), "Should not contain %v", mode)
+			}
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can we use assert.ElementsMatch directly here ?

File: types/stream_test.go [hunk 11]
```diff
@@ -0,0 +1,793 @@
...
+		})
+	}
+}
+
+func TestStream_WithPrimaryKey(t *testing.T) {
+	tests := []struct {
+		name            string
+		keys            []string
+		expectedKeys    []string
+		notExpectedKeys []string
```
Comment (saksham-datazip) [inline types/stream_test.go]: No need for this extra variable notExpectedKeys?

File: types/stream_test.go [hunk 12]
```diff
@@ -0,0 +1,793 @@
...
+
+			for _, key := range tt.expectedKeys {
+				assert.True(t, stream.SourceDefinedPrimaryKey.Exists(key), "Should contain '%s'", key)
+			}
+
+			for _, key := range tt.notExpectedKeys {
+				assert.False(t, stream.SourceDefinedPrimaryKey.Exists(key), "Should not contain '%s'", key)
+			}
+		})
+	}
```
Comment (saksham-datazip) [inline types/stream_test.go:284]: please try here assert.ElementsMatch as well if possible?

File: types/stream_test.go [hunk 13]
```diff
@@ -0,0 +1,793 @@
...
+}
+
+func TestStream_WithCursorField(t *testing.T) {
+	tests := []struct {
+		name              string
+		fields            []string
+		expectedFields    []string
+		notExpectedFields []string
```
Comment (saksham-datazip) [inline types/stream_test.go]: Just make the respective changes of removing this extra variable from every other unit tests as welll please?

File: types/stream_test.go [hunk 14]
```diff
@@ -0,0 +1,793 @@
...
+
+		stream.WithSchema(firstSchema)
+		assert.Same(t, firstSchema, stream.Schema, "First schema should be set")
+
+		returnedStream := stream.WithSchema(secondSchema)
+		assert.Same(t, stream, returnedStream, "Should return the same stream instance")
+		assert.Same(t, secondSchema, stream.Schema, "Second schema should replace first schema")
+		assert.NotSame(t, firstSchema, stream.Schema, "First schema should no longer be referenced")
+	})
+}
```
Comment (saksham-datazip) [inline types/stream_test.go]: Like the way we made test struct for other unit tests we can do the same here as well and test cases will look something like this and if you have a better solution than feel free to discuss.
```
{
			name:           "set schema",
			schema:         setSchema,
			expectedSchema: setSchema,
		},
		{
			name:           "clear schema",
			schema:         nil,
			expectedSchema: nil,
		},

File: types/stream_test.go [hunk 15]
```diff
@@ -0,0 +1,793 @@
...
+
+func TestStream_Wrap(t *testing.T) {
```
Comment (saksham-datazip) [inline types/stream_test.go]: Lets follow the functioning order which is in streams.go so please move this function below upsertField and also check for other places if their are any where such ordering is wrong.

File: types/stream_test.go [hunk 16]
```diff
@@ -0,0 +1,793 @@
...
+		{
+			name:      "wrap with negative index",
+			syncIndex: -1,
+		},
+		{
+			name:      "wrap with large index",
+			syncIndex: 100,
+		},
+	}
+
```
Comment (saksham-datazip) [inline types/stream_test.go:346]: i think the input in this function is being ignored so no need for this much of test cases what do you say ?

File: types/stream_test.go [hunk 17]
```diff
@@ -0,0 +1,793 @@
...
+		})
+	}
+}
+
+func TestStream_WithUpsertField(t *testing.T) {
+	tests := []struct {
+		testName string
+
+		column        string
+		dataType      DataType
```
Comment (saksham-datazip) [inline types/stream_test.go]: change this dataType variable to actual variable name provided in input of this function

File: types/stream_test.go [hunk 18]
```diff
@@ -0,0 +1,793 @@
...
+
+			assert.NotNil(t, configuredStream, "Should return a configuredStream")
+			assert.Same(t, stream, configuredStream.Stream, "Should wrap the exact same stream instance")
+		})
+	}
+}
+
+func TestStream_WithUpsertField(t *testing.T) {
+	tests := []struct {
+		testName string
```
Comment (saksham-datazip) [inline types/stream_test.go:374]: Somewhere you are using testName and in some other parts of code you are using name alone so can we keep variable name consistent as well in this entire file?

File: types/stream_test.go [hunk 19]
```diff
@@ -0,0 +1,793 @@
...
+
+		column        string
+		dataType      DataType
+		nullable      bool
+		isOLakeColumn bool
+
+		expectedTypes    []DataType
+		notExpectedTypes []DataType
```
Comment (saksham-datazip) [inline types/stream_test.go]: remove this notExpectedTypes variable

File: types/stream_test.go [hunk 20]
```diff
@@ -0,0 +1,793 @@
...
+			property := value.(*Property)
+			asserts.Equal(utils.Reformat(tt.column), property.DestinationColumnName)
+			asserts.Equal(tt.expectedOlakeCol, property.OlakeColumn)
+
+			for _, expectedType := range tt.expectedTypes {
+				asserts.True(property.Type.Exists(expectedType), "Type set should have '%v'", expectedType)
+			}
+			for _, notExpectedType := range tt.notExpectedTypes {
+				asserts.False(property.Type.Exists(notExpectedType), "Type set should not have '%v'", notExpectedType)
+			}
```
Comment (saksham-datazip) [inline types/stream_test.go]: This part can be simplified by alot using ElementsMatch ig can u check that once out ?

File: types/stream_test.go [hunk 21]
```diff
@@ -0,0 +1,793 @@
...
+
+			for _, field := range tt.expectedFields {
+				assert.True(t, stream.AvailableCursorFields.Exists(field), "Should contain '%s'", field)
+			}
+
+			for _, field := range tt.notExpectedFields {
+				assert.False(t, stream.AvailableCursorFields.Exists(field), "Should not contain '%s'", field)
+			}
+		})
+	}
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can we simpify this part as well?

File: types/stream_test.go [hunk 22]
```diff
@@ -0,0 +1,793 @@
...
+
+		asserts.True(property.Type.Exists(Int64), "1st assert should be present TYPE : 'INT64'")
+		asserts.True(property.Type.Exists(String), "2nd assert should be present TYPE : 'String'")
+		asserts.True(property.Type.Exists(Null), "2nd assert added Null as well should be present TYPE : 'Null'")
+
+		asserts.Equal(false, property.OlakeColumn)
+	})
+}
+
+func TestStream_UnmarshalJSON(t *testing.T) {
```
Comment (saksham-datazip) [inline types/stream_test.go]: Please try to follow patterns as the way we followed in other files by creating a struct and also make the same changes in function below as well.

**Review by saksham-datazip — COMMENTED** [review]
Sorry for the late review and please make the respective changes.

**Comment by supersanchayrx** [comment]
Hey sure I'll fix these asap thanks for the review

**Comment by saksham-datazip** [comment]
@supersanchayrx Are you still working on this unit test?

File: types/stream_test.go (hunk 21 above)
Comment (supersanchayrx) [reply types/stream_test.go]: removed the not expected loop and used ElementsMatch method

File: types/stream_test.go (hunk 1 above)
Comment (supersanchayrx) [reply types/stream_test.go]: renamed to TestNewStream

Comment (supersanchayrx) [reply types/stream_test.go]: removed all the line changes and whitespaces causing these tests to fail

File: types/stream_test.go (hunk 3 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Makes sense, I removed the viper prefix field and the test case as well

File: types/stream_test.go (hunk 2 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Done. Defined an expectedStream and compared results directly against it.

File: types/stream_test.go (hunk 4 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Done!

File: types/stream_test.go (hunk 6 above)
Comment (supersanchayrx) [reply types/stream_test.go]: renamed to TestStreamID (as in reformat test file with no underscores)

File: types/stream_test.go (hunk 7 above)
Comment (supersanchayrx) [reply types/stream_test.go]: made test names cleaner and comprehensive for easy review

File: types/stream_test.go (hunk 8 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Done! renamed to TestStreamWithSyncMode

File: types/stream_test.go (hunk 9 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Removed and apologies for not double checking as i built it over from the last pr #680 raised for this issue.

File: types/stream_test.go (hunk 10 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Yes done!

File: types/stream_test.go (hunk 11 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Yes since the iteration during insert is only on the keys we've passed oruselves there's no need to check if wrong keys can be added. Removed it and subsequent test cases!

File: types/stream_test.go (hunk 12 above)
Comment (supersanchayrx) [reply types/stream_test.go:284]: Done!

File: types/stream_test.go (hunk 13 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Yup done!

File: types/stream_test.go (hunk 15 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Yes fixed the order

File: types/stream_test.go (hunk 18 above)
Comment (supersanchayrx) [reply types/stream_test.go:374]: fixed
all test identifiers are not testName

File: types/stream_test.go (hunk 17 above)
Comment (supersanchayrx) [reply types/stream_test.go]: done

File: types/stream_test.go (hunk 19 above)
Comment (supersanchayrx) [reply types/stream_test.go]: removed field and subsequent test cases as well

**Comment by supersanchayrx** [comment]
Hey @saksham-datazip sorry for the delay. Yup I implemented your suggestions and fixed the code wherever needed. Please review it once more. Thanks!

File: types/stream_test.go (hunk 16 above)
Comment (saksham-datazip) [reply types/stream_test.go:346]: ?

File: types/stream_test.go (hunk 22 above)
Comment (saksham-datazip) [reply types/stream_test.go]: Could you please use the same test case structure that's used in the other tests as well? It would help maintain consistency across the test suite for All the functions.Like this:-
```
	tests := []struct {
		testName  string
		syncIndex int
	}{
```
Instead of directly using t.Run

File: types/stream_test.go [hunk 23]
```diff
@@ -0,0 +1,717 @@
...
+			stream := NewStream(tt.name, tt.namespace, tt.sourceDatabase)
+
+			asserts.Equal(tt.expectedStream.Name, stream.Name)
+			asserts.Equal(tt.expectedStream.Namespace, stream.Namespace)
+			asserts.NotNil(stream.SupportedSyncModes, "SupportedSyncModes should be initialized")
+			asserts.NotNil(stream.SourceDefinedPrimaryKey, "SourceDefinedPrimaryKey should be initialized")
+			asserts.NotNil(stream.AvailableCursorFields, "AvailableCursorFields should be initialized")
+			asserts.NotNil(stream.Schema, "Schema should be initialized")
+			asserts.Equal(tt.expectedStream.DestinationDatabase, stream.DestinationDatabase)
+			asserts.Equal(tt.expectedStream.DestinationTable, stream.DestinationTable)
```
Comment (saksham-datazip) [inline types/stream_test.go]: instead of this much checks you should use asserts.Equal(tt.expectedStream, stream) 
and apart from this the streams which we are getting as output are different from expected so consider using complete structure, Like this :-
```
			expectedStream: &Stream{
				Name:                    "User-Orders.v2",
				Namespace:               "My.Schema",
				Schema:                  NewTypeSchema(),
				SupportedSyncModes:      NewSet[SyncMode](),
				SourceDefinedPrimaryKey: NewSet[string](),
				AvailableCursorFields:   NewSet[string](),
				DestinationDatabase:     ":my_schema",
				DestinationTable:        "user_orders_v2",
			},
```
**Note:** This is a dummy code

File: types/stream_test.go [hunk 24]
```diff
@@ -0,0 +1,717 @@
...
+			keys:         []string{"user_id", "order_id"},
+			expectedKeys: []string{"user_id", "order_id"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.testName, func(t *testing.T) {
+			stream := NewStream("users", "public", nil)
+			returnedStream := stream.WithPrimaryKey(tt.keys...)
+
```
Comment (saksham-datazip) [inline types/stream_test.go:279]: returnedStream was previously named outputStream. Could we make the naming consistent? Also, if there are any other variable names that could be made more consistent throughout the test for better readability, it would be nice to update those as well.

File: types/stream_test.go [hunk 25]
```diff
@@ -0,0 +1,717 @@
...
+		{
+			testName:     "multiple keys",
+			keys:         []string{"id", "user_uuid"},
+			expectedKeys: []string{"id", "user_uuid"},
+		},
+		{
+			testName:     "composite key",
+			keys:         []string{"tenant_id", "user_id", "order_id"},
+			expectedKeys: []string{"tenant_id", "user_id", "order_id"},
+		},
```
Comment (saksham-datazip) [inline types/stream_test.go]: multiple keys test and composite keys tests are kind of same just with different number of keys and due to their naming ig this testName "keys with underscores" is redundant.

File: types/stream_test.go [hunk 26]
```diff
@@ -0,0 +1,717 @@
...
+			assert.ElementsMatch(t, tt.expectedKeys, stream.SourceDefinedPrimaryKey.Array())
+		})
+	}
+}
+
+func TestStreamWithCursorField(t *testing.T) {
+	tests := []struct {
+		testName       string
+		fields         []string
+		expectedFields []string
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can you change this fields name with the function input name also check for other variables.

File: types/stream_test.go [hunk 27]
```diff
@@ -0,0 +1,717 @@
...
+			fields:         []string{"updated_at", "updated_at", "inserted_at"},
+			expectedFields: []string{"updated_at", "inserted_at"},
+		},
+		{
+			testName:       "empty fields",
+			fields:         []string{},
+			expectedFields: []string{},
+		},
+		{
+			testName:       "fields with underscores",
```
Comment (saksham-datazip) [inline types/stream_test.go]: Also Can you please confirm these all checks seems to be redundant ?

File: types/stream_test.go [hunk 28]
```diff
@@ -0,0 +1,717 @@
...
+
+			assert.Same(t, stream, outputStream, "Should return the same instance")
+
+			assert.ElementsMatch(t, tt.expectedFields, stream.AvailableCursorFields.Array())
+		})
+	}
+}
+
+func TestStreamWithSchema(t *testing.T) {
+	t.Run("set new schema", func(t *testing.T) {
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can you please use struct using testcases instead of directly calling t.Run and also for functions TestStreamUnmarshalJSON, TestStreamsToMap, TestLogCatalog do the same.

File: types/stream_test.go [hunk 29]
```diff
@@ -0,0 +1,717 @@
...
+
+	err = json.Unmarshal(content, &savedCatalog)
+	assert.NoError(t, err, "File content should be valid JSON")
+	assert.Equal(t, 2, len(savedCatalog.Streams), "Saved catalog should contain 2 streams")
+}
+
+// helper to get string pointer
+func stringPtr(s string) *string {
+	return &s
+}
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can you inline this function no need for a new function?

File: types/stream_test.go [hunk 30]
```diff
@@ -0,0 +1,717 @@
...
+	}
+
+	//test for merging two datatypes on subsequent upsert calls with same names
+	t.Run("Multiple datatypes test", func(t *testing.T) {
+		asserts := assert.New(t)
+		stream := NewStream("phones", "seller", nil)
+
+		stream.UpsertField("codename", Int64, false, false)
+		stream.UpsertField("codename", String, true, true)
+
```
Comment (saksham-datazip) [inline types/stream_test.go:507]: Can we write this testcase like others no need for a new t.Run ?

File: types/stream_test.go [hunk 31]
```diff
@@ -0,0 +1,717 @@
...
+			syncIndex: 1,
+		},
+		{
+			testName:  "wrap with negative index",
+			syncIndex: -1,
+		},
+		{
+			testName:  "wrap with large index",
+			syncIndex: 100,
+		},
```
Comment (saksham-datazip) [inline types/stream_test.go]: Look into those tc as this function is input independent 1-2 basic testcases is more than enough

File: types/stream_test.go [hunk 32]
```diff
@@ -0,0 +1,717 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.testName, func(t *testing.T) {
+			stream := NewStream("users", "public", nil)
+			configuredStream := stream.Wrap(tt.syncIndex)
```
Comment (saksham-datazip) [inline types/stream_test.go]: here also try to make this name configuredStream consistent as told earlier

File: types/stream_test.go [hunk 33]
```diff
@@ -0,0 +1,717 @@
...
+
+			assert.NotNil(t, configuredStream, "Should return a configuredStream")
+			assert.Same(t, stream, configuredStream.Stream, "Should wrap the exact same stream instance")
```
Comment (saksham-datazip) [inline types/stream_test.go]: Can we use equal instead of those two lines ?

File: types/stream_test.go [hunk 34]
```diff
@@ -0,0 +1,717 @@
...
+		})
+	}
+}
+
```
Comment (saksham-datazip) [inline types/stream_test.go:545]: For the function after line 538 i am expecting you to make changes if possible else a reason why we cant follow olake pattern.

**Review by saksham-datazip — COMMENTED** [review]
I've addressed the previous comments on my end, but it looks like a few comments are still pending from your side. Could you please take a look at those as well? If you have any questions or need any clarification, feel free to ask in the comments.

**Comment by saksham-datazip** [comment]
@supersanchayrx Hey! Are you still working on this issue? If so, please let us know. Otherwise, we'll have to close this PR due to inactivity.

**Comment by supersanchayrx** [comment]
hey yes sorry for the delay. I'm pushing an update asap

File: types/stream_test.go (hunk 23 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Done! thanks for the dummy code 👍

File: types/stream_test.go (hunk 18 above)
Comment (supersanchayrx) [reply types/stream_test.go:374]: made the variable name consistent like other test files ('tt' -> 'tc')

File: types/stream_test.go (hunk 25 above)
Comment (supersanchayrx) [reply types/stream_test.go]: Yes i've removed the redundant test cases.

File: types/stream_test.go (hunk 26 above)
Comment (supersanchayrx) [reply types/stream_test.go]: changed 'fields' -> 'columns' to match the function input var name

File: types/stream_test.go (hunk 27 above)
Comment (supersanchayrx) [reply types/stream_test.go]: yes removed the redundant check

File: types/stream_test.go (hunk 28 above)
Comment (supersanchayrx) [reply types/stream_test.go]: yes modified it into a struct based test. I just kept the replace case differently as it doesn't fit into this struct tests' input type.

File: types/stream_test.go (hunk 30 above)
Comment (supersanchayrx) [reply types/stream_test.go:507]: I can do it but I purposely kept it separate as it needs to check whether 2 usertfields on 1 col actually gives merged types. Making this into the table tests would reduce readibilty in my opinion and would also look a bit odd but nonetheless I can make it happen. Do let me know I'll make that change quick

File: types/stream_test.go (hunk 31 above)
Comment (supersanchayrx) [reply types/stream_test.go]: yeah removed trivial ones. Just kept the -ve and 0

File: types/stream_test.go (hunk 24 above)
Comment (supersanchayrx) [reply types/stream_test.go:279]: I have changed this and also tried to keep consistent naming across file

File: types/stream_test.go (hunk 22 above)
Comment (supersanchayrx) [reply types/stream_test.go]: done
it now uses this kind of tests struct pattern only. it has per field expected values to cross check against. all of the test cases are traversed through loop. Made sure this pattern remains consistent with other methods as well.

File: types/stream_test.go (hunk 34 above)
Comment (supersanchayrx) [reply types/stream_test.go:545]: done. Sorry for not resolving this in last review
it now uses this kind of tests struct pattern only as you suggested in other comment. it has per field expected values to cross check against. all of the test cases are traversed through loop. Made sure this pattern remains consistent with other methods as well.

File: types/stream_test.go (hunk 29 above)
Comment (supersanchayrx) [reply types/stream_test.go]: actually i was going for inline method but it just meant i had to declare a local var at each of the places its called as i couldn't fetch address of string literal inline. Plus I saw helper method in other files as well so I thought its alright. Let me know if not i'll replace it to inline.

File: types/stream_test.go (hunk 33 above)
Comment (supersanchayrx) [reply types/stream_test.go]: yes used equal directly of 2 individual checks.

File: types/stream_test.go (hunk 5 above)
Comment (supersanchayrx) [reply types/stream_test.go]: done. COllapsed the entire multiple checks block to a single equal check.

File: types/stream_test.go (hunk 16 above)
Comment (supersanchayrx) [reply types/stream_test.go:346]: yes figured that out late. i removed other test cases. This also now uses just a single check like suggested for functions earlier

File: types/stream_test.go (hunk 14 above)
Comment (supersanchayrx) [reply types/stream_test.go]: done. I made it follow the same pattern as this dummy code. I

File: types/stream_test.go (hunk 32 above)
Comment (supersanchayrx) [reply types/stream_test.go]: yes followed consistency for all variables in general throughout the file

**Comment by supersanchayrx** [comment]
Please have a review once more

File: types/stream_test.go (hunk 20 above)
Comment (supersanchayrx) [reply types/stream_test.go]: used elementsmatch for the check.

File: types/stream_test.go [hunk 35]
```diff
@@ -0,0 +1,751 @@
...
+func TestStreamID(t *testing.T) {
+	tests := []struct {
+		testName   string
+		name       string
+		namespace  string
+		expectedID string
```
Comment (saksham-datazip) [inline types/stream_test.go:155]: I already commented right that we can return expectedStream instead of this as we treat unit test as a black box just gave a input and check hardcoded output ?And same thing applies to rest of the function as well

File: types/stream_test.go [hunk 36]
```diff
@@ -0,0 +1,751 @@
...
+
+			assert.Same(t, stream, returnedStream, "should return the same stream instance")
+			assert.Same(t, tc.schema, stream.Schema, "Stream.Schema should point to the expected schema")
+		})
+	}
+
+	t.Run("replace existing schema", func(t *testing.T) {
+		stream := NewStream("users", "public", nil)
+		firstSchema := NewTypeSchema()
+		secondSchema := NewTypeSchema()
```
Comment (saksham-datazip) [inline types/stream_test.go:360]: here i dont think we need a saperate t.Run block and we could have used simple struct format if output was expectedStream ?

File: types/stream_test.go [hunk 37]
```diff
@@ -0,0 +1,751 @@
...
+			asserts := assert.New(t)
+			stream := NewStream("grades", "students", nil)
+			stream.UpsertField(tc.column, tc.typ, tc.nullable, tc.isOLakeColumn)
+
+			value, ok := stream.Schema.Properties.Load(tc.column)
+			asserts.True(ok, "Schema should have column '%s'", tc.column)
+			property := value.(*Property)
+			asserts.Equal(utils.Reformat(tc.column), property.DestinationColumnName)
+			asserts.Equal(tc.expectedOlakeCol, property.OlakeColumn)
+
```
Comment (saksham-datazip) [inline types/stream_test.go:496]: If we had used an expectedStream here, we could have avoided several of these individual assertions and kept the test consistent with the existing OLake test structure.

File: types/stream_test.go (hunk 30 above)
Comment (saksham-datazip) [reply types/stream_test.go:507]: yes and if you had expectedStream instead  to detect output than we could have avoided this .

File: types/stream_test.go [hunk 38]
```diff
@@ -0,0 +1,751 @@
...
+		testName            string
+		jsonData            []byte
+		wantErr             bool
+		expectedName        string
+		expectedNamespace   string
+		expectedSyncMode    SyncMode
+		expectedCursorField string
+		expectedSyncModes   []SyncMode
+		expectedPrimaryKeys []string
+		expectedCursors     []string
```
Comment (saksham-datazip) [inline types/stream_test.go:557]: ExpectedStreams ?

File: types/stream_test.go [hunk 39]
```diff
@@ -0,0 +1,751 @@
...
+		})
+	}
+}
+
+func TestStreamUnmarshalJSON(t *testing.T) {
+	tests := []struct {
+		testName            string
+		jsonData            []byte
+		wantErr             bool
+		expectedName        string
```
Comment (saksham-datazip) [inline types/stream_test.go:551]: and why we do we have this wantErr you can check in reformat_test.go that we directly compare Err not just check it.

**Review by saksham-datazip — COMMENTED** [review]
Third Review

**Comment by saksham-datazip** [comment]
Closing this PR for now, as the changes seem to be AI generated and will need some more review and discussion before we can approve them.
